Updated to python3

alisonrclarke opened the pull request:
Fixes #30

GraemeWatt commented on hepdata_converter/parsers/oldhepdata_parser.py:
```diff
@@ -126,7 +127,7 @@ def parse(self, data_in):
         self.reset()
         # in case of strings we should treat them as filepaths
-        if isinstance(data_in, (str, unicode)):
+        if isinstance(data_in, (str, str)):
```
`if isinstance(data_in, str):`

GraemeWatt commented on hepdata_converter/testsuite/__init__.py:
```diff
@@ -98,16 +101,16 @@ class ExtendedTestCase(unittest.TestCase):
         if hasattr(first, 'readlines'):
             lines = first.readlines()
-        elif isinstance(first, (str, unicode)):
+        elif isinstance(first, (str, str)):
```
`elif isinstance(first, str):`

GraemeWatt commented on hepdata_converter/testsuite/__init__.py:
```diff
@@ -98,16 +101,16 @@ class ExtendedTestCase(unittest.TestCase):
         if hasattr(second, 'readlines'):
             orig_lines = second.readlines()
-        elif isinstance(second, (str, unicode)):
+        elif isinstance(second, (str, str)):
```
`elif isinstance(second, str):`

GraemeWatt commented on hepdata_converter/writers/array_writer.py:
```diff
@@ -21,7 +24,7 @@ def is_number_var(cls, *variables):
         for variable in variables:
             for element in variable['values']:
-                if 'value' in element and isinstance(element['value'], (str, unicode)):
+                if 'value' in element and isinstance(element['value'], (str, str)):
```
`if 'value' in element and isinstance(element['value'], str):`

GraemeWatt commented on hepdata_converter/writers/array_writer.py:
```diff
@@ -265,7 +266,7 @@ def _prepare_outputs(self, data_out, outputs):
                     outputs.append(open(os.path.join(data_out, table.name.replace(' ','').replace('/','-').replace('$','').replace('\\','') + '.' + self.extension), 'w'))
         # multiple tables - require directory
-        elif len(self.tables) > 1 and not isinstance(data_out, (str, unicode)):
+        elif len(self.tables) > 1 and not isinstance(data_out, (str, str)):
```
`elif len(self.tables) > 1 and not isinstance(data_out, str):`

GraemeWatt commented on hepdata_converter/writers/array_writer.py:
```diff
@@ -367,19 +368,19 @@ def _parse_dependent_variable(cls, dependent_variable, headers, qualifiers, qual
                         if 'symerror' in error:
                             error_plus = error['symerror']
-                            if isinstance(error_plus, (str, unicode)):
+                            if isinstance(error_plus, (str, str)):
```
`if isinstance(error_plus, str):`

GraemeWatt commented on hepdata_converter/writers/root_writer.py:
```diff
@@ -419,7 +421,7 @@ def _prepare_outputs(self, data_out, outputs):
         """
         compress = ROOTModule.ROOT.CompressionSettings(ROOTModule.ROOT.kZLIB, 1)
-        if isinstance(data_out, (str, unicode)):
+        if isinstance(data_out, (str, str)):
```
`if isinstance(data_out, str):`

GraemeWatt commented on hepdata_converter/writers/utils.py:
```diff
@@ -8,11 +10,11 @@ def error_value_processor(value, error):
     :return: the absolute error, e.g. 12 for the above case.
     """
-    if isinstance(error, (str, unicode)):
+    if isinstance(error, (str, str)):
```
`if isinstance(error, str):`

GraemeWatt commented on hepdata_converter/writers/yaml_writer.py:
```diff
@@ -48,7 +49,7 @@ def write(self, data_in, data_out, *args, **kwargs):
                 table.metadata['table_doi'] = self.hepdata_doi + '/t' + str(table.index)
 
-        if not isinstance(data_out, (str, unicode)) and not self.single_file:
+        if not isinstance(data_out, (str, str)) and not self.single_file:
```
`if not isinstance(data_out, str) and not self.single_file:`

GraemeWatt commented on hepdata_converter/writers/yaml_writer.py:
```diff
@@ -60,7 +61,7 @@ def write(self, data_in, data_out, *args, **kwargs):
                         yaml.dump(table.data, table_file, Dumper=Dumper, default_flow_style=None)
         else:
-            if isinstance(data_out, (str, unicode)):
+            if isinstance(data_out, (str, str)):
```
`if isinstance(data_out, str):`

GraemeWatt commented on hepdata_converter/writers/yoda_writer.py:
```diff
@@ -92,7 +94,7 @@ def __init__(self, *args, **kwargs):
 
     def _prepare_outputs(self, data_out, outputs):
-        if isinstance(data_out, (str, unicode)):
+        if isinstance(data_out, (str, str)):
```
`if isinstance(data_out, str):`

GraemeWatt commented on hepdata_converter/writers/yoda_writer.py:
```diff
@@ -120,10 +122,8 @@ def _write_table(self, data_out, table):
-#            graph.setTitle(table_doi)  # use for YODA 1.8.1
-#            graph.setPath(rivet_path)  # use for YODA 1.8.1
+            graph.setTitle(table_doi)  # use for YODA 1.8.1
+            graph.setPath(rivet_path)  # use for YODA 1.8.1
```
Change comment to `# use for YODA > 1.8.0`.

alisonrclarke replied:
Should I just delete the comments now we don't have two versions of the code?

GraemeWatt replied:
Yes, better just to delete the comment, since we won't downgrade YODA in future.

GraemeWatt commented on hepdata_converter/writers/utils.py:
```diff
@@ -8,11 +10,11 @@ def error_value_processor(value, error):
             if "%" in error:
                 error_float = float(error.replace("%", ""))
-                error_abs = (value/100) * error_float
+                error_abs = (old_div(value,100)) * error_float
```
Previous line was fine in Python 3, i.e. `(value/100) * error_float`.

alisonrclarke commented:
> The supported Python version (3.6 or later) should also be updated in `travis.yml`, `.readthedocs.yml`, and `setup.py`.

I'll change to python 3.6 in `.travis.yml` for consistency but it's worth noting that all the python code in the travis script is run via docker, so it shouldn't affect anything. On that point, do we need the install steps at https://github.com/DurhamARC/hepdata-converter/blob/e173022edbfe521726212c5059662c1e6975b16b/.travis.yml#L18 ?

GraemeWatt commented:
> I'll change to python 3.6 in `.travis.yml` for consistency but it's worth noting that all the python code in the travis script is run via docker, so it shouldn't affect anything. On that point, do we need the install steps at https://github.com/DurhamARC/hepdata-converter/blob/e173022edbfe521726212c5059662c1e6975b16b/.travis.yml#L18 ?

Probably Python 3.7 is better to match the version used in the `hepdata-converter` Docker image.  I think it's only needed for running `coveralls`, i.e. push the `coverage` output to [coveralls.io](https://coveralls.io/github/HEPData/hepdata-converter?branch=master).  Probably the line `sudo pip install coverage` can be removed since `coverage` will be installed by `coveralls`.  It might not be ideal that `coverage` runs inside the Docker container, but `coveralls` runs outside, but it seems to work and I don't know if `coveralls` would work inside the Docker container.

alisonrclarke commented:
I think I've addressed all the comments now, but before this can be merged we need to ensure hepdata/hepdata-converter-docker#6 is OK, build a new docker image, and update both the docker image tag and the docker submodule here.

GraemeWatt commented:
I've merged HEPData/hepdata-converter-docker/pull/6 and the updated `hepdata/hepdata-converter:0.2.0` image is available on [Docker Hub](https://hub.docker.com/r/hepdata/hepdata-converter).  Could you please update the `docker` submodule before I merge?

GraemeWatt commented:
I had to restart the [Travis CI job](https://travis-ci.org/github/HEPData/hepdata-converter/builds/703218183) since there were some temporary problems this afternoon.  From the job log, it looks like `sudo pip install` uses Python 2.7 to install packages globally rather than installing for Python 3.7 in a virtual environment.  This causes problems at the end of the job when running `coveralls`.  Could you please remove the `sudo` from the `pip install` lines and also the deprecated `sudo: required` line from the `.travis.yml` file?
